src/carousel.tsx

```diff
@@ -0,0 +1,727 @@
+    const [itemScrollPositions, setItemScrollPositions] = React.useState<Array<number>>([]);
+
+    const pagesScrollPositions = calcPagesScrollPositions(itemScrollPositions, pagesCount);
+    const scrollPositions = itemsToScroll
+        ? calcPagesScrollPositions(itemScrollPositions, Math.ceil(items.length / itemsToScroll))
+        : pagesScrollPositions;
+
+    const showNextArrow = scrollRight !== 0;
+    const showPrevArrow = scrollLeft !== 0;
+
+    React.useLayoutEffect(() => {
+        if (carouselRef.current) {
+            const carouselEl = carouselRef.current;
+
+            const handleCarouselChange = () => {
+                const {scrollWidth, clientWidth} = carouselEl;
+                const scrollLeft = Math.round(carouselEl.scrollLeft);
+
+                const scrollRight = Math.round(scrollWidth - (scrollLeft + clientWidth));
+
+                setScroll({scrollLeft, scrollRight});
+            };
+
+            const calcItemScrollPositions = () => {
+                const maxScroll = carouselEl.scrollWidth - carouselEl.clientWidth;
+
+                setItemScrollPositions(
+                    Array.from(carouselEl.querySelectorAll('[data-item]')).map((itemEl, idx) => {
+                        if (idx === items.length - 1) {
+                            return maxScroll;
+                        }
+                        const offsetLeft = (itemEl as HTMLElement).offsetLeft;
+                        const scrollMargin = Number(getComputedStyle(itemEl).scrollMargin.replace('px', ''));
+                        const scrollPosition =
+                            centered && !isDesktopOrBigger ? offsetLeft - itemEl.clientWidth / 2 : offsetLeft;
+                        return Math.min(scrollPosition - scrollMargin - carouselEl.offsetLeft, maxScroll);
+                    })
+                );
+            };
+
+            handleCarouselChange();
+            calcItemScrollPositions();
+
+            carouselEl.addEventListener('scroll', handleCarouselChange);
+            const unlistenResize = listenResize(carouselEl, () => {
+                handleCarouselChange();
+                calcItemScrollPositions();
+            });
+
+            return () => {
+                carouselEl.removeEventListener('scroll', handleCarouselChange);
+                unlistenResize();
+            };
+        }
+        return () => {};
+    }, [
+        itemsPerPageConfig.desktop,
+        itemsPerPageConfig.mobile,
+        mobilePageOffsetConfig.next,
+        mobilePageOffsetConfig.prev,
+        pagesCount,
+        gap,
+        centered,
+        isDesktopOrBigger,
+        sideMargin,
+        items.length,
+    ]);
+
+    const goToPage = React.useCallback(
+        (pageIndex: number) => {
+            const carouselEl = carouselRef.current;
+            if (carouselEl) {
+                const scroll = pagesScrollPositions[pageIndex];
+                carouselEl.scrollTo({left: scroll, behavior: 'smooth'});
```

Comment:
does this work in ios? it doesn't in old ios

Reply:
You're right, old iOS don't support the smooth option, but the percentage of old iOS versions is low and the graceful degradation is good enough. Also, arrow buttons are not used in mobile, so not a big deal